packages/react-chain/src/ReactChain.ts

```diff
@@ -1,108 +1,78 @@
-    Promise<ReactElement<any>>
+  (renderChildren: () => Promise<null | ReactElement<any>>, context: any) =>
+    ReactElement<any> | Promise<ReactElement<any>>
 
 export type WrapRender =
-  (render: Function, context: Context<any>) =>
+  (render: Function) =>
     void
 
-export interface Middleware {
-  createElement?: WrapElement
-  wrapClientRender?: WrapRender
-  wrapServerRender?: WrapRender
-}
+export type Middleware =
+  (session: Session) =>
+    (void | WrapElement)
 
 export class ReactChain {
+  protected middlewareChain: Array<Middleware> = []
   protected elementChain: Array<WrapElement> = []
-  protected clientChain: Array<WrapRender> = []
-  protected serverChain: Array<WrapRender> = []
-
-  chain(middleware: WrapElement | Middleware) {
-    if (typeof middleware === 'function') {
-      middleware = {
-        createElement: middleware,
-      }
-    }
 
-    if (middleware.createElement) {
-      if (typeof middleware.createElement !== 'function') {
-        throw new Error('[chain()] createElement should be a function.')
-      }
-      this.elementChain.push(middleware.createElement)
-    }
+  private firstRender = true
 
-    if (middleware.wrapClientRender) {
-      if (typeof middleware.wrapClientRender !== 'function') {
-        throw new Error('[chain()] wrapClientRender should be a function.')
-      }
-      this.clientChain.push(middleware.wrapClientRender)
+  chain(middleware?: Middleware) {
+    if (typeof middleware !== 'function') {
+      throw new Error('A react-chain middleware should be a function')
     }
 
-    if (middleware.wrapServerRender) {
-      if (typeof middleware.wrapServerRender !== 'function') {
-        throw new Error('[chain()] wrapServerRender should be a function.')
-      }
-      this.serverChain.push(middleware.wrapServerRender)
-    }
+    this.middlewareChain.push(middleware)
+    this.firstRender = true
 
     return this
   }
 
-  async getElement(context = createContext({})): Promise<ReactElement<any>> {
-    const elementChain = [createBase, ...this.elementChain]
+  async getElement(session = new Session()): Promise<ReactElement<any>> {
+    if (this.firstRender) {
+      this.elementChain = [createBase]
+
+      for (const middleware of this.middlewareChain) {
+        const createElement = middleware(session)
+
+        if (typeof createElement === 'function') {
+          this.elementChain.push(createElement)
+        }
+      }
+    }
+
     let index = 0
+    this.firstRender = false
 
-    async function next(): Promise<any> {
-      const createElement = elementChain[index++]
-      const renderChildren = elementChain[index]
+    const next = async (): Promise<any> => {
+      const createElement = this.elementChain[index++]
+      const renderChildren = this.elementChain[index]
         ? await next
         : () => Promise.resolve(null)
-      return createElement(renderChildren, context)
+      return createElement(renderChildren, session)
```

Comment:
This async/promise function feels weird to me. It shouldn't need to be async, no need to `await next`, (a function, not a promise). Everything otherwise returns a promise.

Except we should probably do `Promise.resolve(createElement(...))`.